en/dev_log/log_encryption.md

```diff
@@ -0,0 +1,233 @@
+
+In order to extract the log file, a user must first decrypt the wrapped symmetric key, which can then be used to decrypt the log.
+Note that decrypting the symmetric key file is only possible if the user has the appropriate RSA private key (corresponding to the public key that was used to wrap it).
+This process is covered in [Download & Decrypt Log Files](#download-decrypt-log-files) below.
+
+## Custom PX4 Firmware with Log Encryption
+
+You will need to build custom firmware that contains your own public RSA key and the required Crypto API modules to support log encryption.
+This section shows how to do this using the `px4-fmu-v5` board as an example.
+
+::: tip
+We show you how to generate your own keys in the [Generate RSA Public & Private Keys](#generate-rsa-public-private-keys) section below.
+:::
+
+::: info
+The modules in a PX4 build are defined in configuration files, which may be modified either manually or using the `menuconfig` tool.
+For more information see: [PX4 Board Configuration (Kconfig)](../hardware/porting_guide_config.md).
+:::
+
+### Cryptotest Make Target
+
+Crypto uses large amounts of flash memory, and is therefore not included in the default PX4 make targets for each board (such as `make px4-fmu-v5`).
+The easiest way to add support for encrypted logs is to define a custom `make` target that includes the required modules and your public RSA keys.
+
+::: warning
+Crypto uses a lot of flash memory, and many builds are close to their maximum capacity.
+If you run into a build error telling you that you have gone above the maximum flash memory, you will need to disable other features in the `.px4board` file you are working on, or in the `default.px4board` file.
+Be careful not to disable something you need.
+
+For example, if you found you were running out of memory on FMUv4 boards you could disable SIH mode by setting `CONFIG_MODULES_SIMULATION_SIMULATOR_SIH=n` in [boards/px4/fmu-v4/default.px4board](https://github.com/PX4/PX4-Autopilot/blob/main/boards/px4/fmu-v4/default.px4board#L76), which may free up enough flash memory to allow crypto to be added.
+:::
+
+#### Pixhawk FMUv5 boards
+
+The FMUv5 board already has a custom make target `px4-fmu-v5_cryptotest` that you can use to build custom firmware with the required modules and "test" RSA keys.
+The configuration file that enables the above make target is [`cryptotest.px4board`](https://github.com/PX4/PX4-Autopilot/blob/main/boards/px4/fmu-v5/cryptotest.px4board) file in `boards/px4/fmu-v5`.
+The relevant keys in that file are reproduced below:
+
+```config
+CONFIG_BOARD_CRYPTO=y
+CONFIG_DRIVERS_STUB_KEYSTORE=y
+CONFIG_DRIVERS_SW_CRYPTO=y
+CONFIG_PUBLIC_KEY1="../../../Tools/test_keys/rsa2048.pub"
+```
+
+::: info
+The file also sets `CONFIG_PUBLIC_KEY0` to a key named `key0.pub`.
+This is not used in the current PX4 implementation and can be ignored.
+:::
+
+::: details Overview of crypto-relevant keys
```

Comment:
FYI @jnomikos I accepted all your changes.
However I have extended this bit to explain how the keys work "according to my understanding".

Essentially the board config copies the keys and required drivers to access them into firmware. But the actual keystore lives in software, so it is seeded by the values in the board config, but after that can modify them.

Not sure if this is also true for hardware-keystore - it might be, but then maybe they also store volatile key data.

Reply:
PS I think this is good to go now otherwise. Just retesting if I can add the values to the config rather than via menuconfig.

Reply:
@jlaitine I'm still stuck with setting the board config value. I think that perhaps they are being set, but I am missing some keys. 

I added the following config to the fmuv4 tree and rebuilt (`>` shows changes to config)

```
> CONFIG_MODULES_SIMULATION_SIMULATOR_SIH=n
CONFIG_MODULES_ATTITUDE_ESTIMATOR_Q=n
CONFIG_MODULES_FW_AUTOTUNE_ATTITUDE_CONTROL=n
CONFIG_MODULES_MC_AUTOTUNE_ATTITUDE_CONTROL=n
> CONFIG_CRYPTO=y
> CONFIG_CRYPTO_RANDOM_POOL=y
> CONFIG_CRYPTO_BLAKE2S
CONFIG_DRIVERS_STUB_KEYSTORE=y
CONFIG_BOARD_CRYPTO=y
CONFIG_DRIVERS_SW_CRYPTO=y
CONFIG_PUBLIC_KEY0="../../../Tools/test_keys/key0.pub"
CONFIG_PUBLIC_KEY1="../../../Tools/test_keys/rsa2048.pub"
```

But I get the error below. I would have thought that undefined reference to `px4_get_secure_random' due to `CONFIG_CRYPTO_RANDOM_POOL` being missing, but I added it above. Can you advise on how to debug?

```
rivers/stub_keystore/libkeystore_backend.a  src/lib/crypto/libmonocypher.a  src/lib/crypto/liblibtomcrypt.a  platforms/nuttx/src/px4/common/libpx4_random.a  NuttX/nuttx/crypto/libcrypto.a  src/lib/crypto/liblibtommath.a  platforms/common/uORB/libuORB.a  NuttX/nuttx/mm/libmm.a  src/lib/cdev/libcdev.a  NuttX/nuttx/fs/libfs.a  msg/libuorb_msgs.a  src/lib/heatshrink/libheatshrink.a  platforms/common/px4_work_queue/libpx4_work_queue.a  NuttX/nuttx/libs/libxx/libxx.a  NuttX/nuttx/drivers/libdrivers.a  NuttX/nuttx/libs/libc/libc.a  NuttX/nuttx/drivers/libdrivers.a  NuttX/nuttx/libs/libc/libc.a  NuttX/nuttx/sched/libsched.a  -lm  -lgcc && :
Memory region         Used Size  Region Size  %age Used
           flash:     1982641 B      2032 KB     95.28%
            sram:       32972 B       192 KB     16.77%
          ccsram:          0 GB        64 KB      0.00/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/drivers/sw_crypto/libcrypto_backend.a(crypto.c.obj): in function `crypto_open':
/home/ubuntu/github/PX4-Autopilot/src/drivers/sw_crypto/crypto.c:188: undefined reference to `px4_get_secure_random'
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/drivers/sw_crypto/libcrypto_backend.a(crypto.c.obj): in function `crypto_generate_key':
/home/ubuntu/github/PX4-Autopilot/src/drivers/sw_crypto/crypto.c:345: undefined reference to `px4_get_secure_random'
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/lib/crypto/liblibtomcrypt.a(libtomcrypt_wrappers.c.obj): in function `rng_get_bytes':
/home/ubuntu/github/PX4-Autopilot/src/lib/crypto/libtomcrypt_wrappers.c:41: undefined reference to `px4_get_secure_random'
collect2: error: ld returned 1 exit status
```

Also, my understanding is that changes made in menuconfig UI are local - i.e. they don't affect any files saved to git right?